pkg/pool/pool.go

```diff
@@ -40,36 +46,41 @@ func NewBytesPool(minSize, maxSize int, factor float64) *BytesPool {
 		sizes = append(sizes, s)
 	}
 
-	p := &BytesPool{
-		buckets: make([]sync.Pool, len(sizes)),
-		sizes:   sizes,
+	p := &Pool{
+		buckets:    make([]sync.Pool, len(sizes)),
+		sizes:      sizes,
+		initialize: newFunc,
 	}
 
 	return p
 }
 
 // Get returns a new byte slices that fits the given size.
-func (p *BytesPool) Get(sz int) []byte {
+func (p *Pool) Get(sz int) interface{} {
 	for i, bktSize := range p.sizes {
 		if sz > bktSize {
 			continue
 		}
-		b, ok := p.buckets[i].Get().([]byte)
-		if !ok {
-			b = make([]byte, 0, bktSize)
+		b := p.buckets[i].Get()
+		if b == nil {
+			b = p.initialize(bktSize)
 		}
 		return b
 	}
-	return make([]byte, 0, sz)
+	return p.initialize(sz)
 }
 
-// Put returns a byte slice to the right bucket in the pool.
-func (p *BytesPool) Put(b []byte) {
-	for i, bktSize := range p.sizes {
-		if cap(b) > bktSize {
-			continue
+// Put adds a slice to the right bucket in the pool.
+func (p *Pool) Put(s interface{}) error {
+	slice := reflect.ValueOf(s)
+	if slice.Kind() == reflect.Slice {
+		for i, size := range p.sizes {
+			if slice.Cap() > size {
+				continue
+			}
+			p.buckets[i].Put(slice.Slice(0, 0).Interface())
+			return nil
 		}
-		p.buckets[i].Put(b[:0])
-		return
 	}
+	return fmt.Errorf("%+v is not a slice", slice)
```

Comment:
I think an error on `Put` is quite annoying to deal with which is in every case a programming bug. Panicking here seems fine and in line with similar things in the standard library, e.g. https://godoc.org/sort#Slice

Also, the shorter conditions should generally be checked first so we can exit early and keep indentations low, e.g.:

```go
slice := reflect.ValueOf(s)

if slice.Kind() != reflect.Slice {
  panic("exit early")
}

for i, size := range p.sizes {
  if slice.Cap() > size {
    continue
  }
  p.buckets[i].Put(slice.Slice(0, 0).Interface())
  return
}
```

Reply:
Probably the check is not necessary as `slice.Cap` will panick on non-slices anyway. But not sure.

Reply:
I thought about panic-ing as well, but wasn't completely sure.
All updated.
Thanks